src/SwarmFit/SwarmFitter.cs

```diff
@@ -57,88 +64,90 @@ double GetError(double[] vars)
     public FitSolution Solve(int iterations = 1000)
     {
         Stopwatch sw = Stopwatch.StartNew();
+        Random rand = Rand;
 
         double[] bestGlobalPositions = new double[VariableCount];
         double bestGlobalError = double.MaxValue;
 
-        Particle[] particles = new Particle[NumParticles];
-        List<FitSolution> intermediateSolutions = [];
+        Span<Particle> particles = new Particle[NumParticles];
+        List<FitSolution>? intermediateSolutions = StoreIntermediateSolutions ? [] : null;
 
         for (int i = 0; i < particles.Length; i++)
         {
-            double[] randomPositions = VarLimits.Select(x => x.Random(Rand)).ToArray();
+            double[] randomPositions = VarLimits.Select(x => x.Random(rand)).ToArray();
             double error = GetError(randomPositions);
-            double[] randomVelocities = VarLimits.Select(x => x.Random(Rand) * VelocityRandomness).ToArray();
+            double[] randomVelocities = VarLimits.Select(x => x.Random(rand) * VelocityRandomness).ToArray();
             particles[i] = new Particle(randomPositions, error, randomVelocities, randomPositions, error);
 
             if (particles[i].Error < bestGlobalError)
             {
                 bestGlobalError = particles[i].Error;
-                particles[i].Positions.CopyTo(bestGlobalPositions, 0);
+                particles[i].Positions.AsSpan().CopyTo(bestGlobalPositions);
             }
         }
 
-        if (StoreIntermediateSolutions)
-        {
-            intermediateSolutions.Add(new(bestGlobalPositions, bestGlobalError, sw.Elapsed, 0, particles.Length));
-        }
+        intermediateSolutions?.Add(new(bestGlobalPositions, bestGlobalError, sw.Elapsed, 0, particles.Length));
+
+        double[] newVelocity = new double[VariableCount];
+        double[] newPosition = new double[VariableCount];
```

Comment:
Both of these arrays are loop-iteration-invariant: they are always overwritten, their length never changes and their references never escape loop body. Therefore, it is optimal to hoist them instead of allocating on each iteration. This change is responsible for the main allocation reduction.